File: Ordination/Python/MergeSort.py
#coding: utf-8

from Sort import Sort

class MergeSort(Sort):
	pass
	
	def sort(self,array,leftIndex = 0 , rightIndex= None):
		if(rightIndex == None):
			rightIndex = len(array) - 1
		if self.__validateParams(array,leftIndex,rightIndex):
			mid = (leftIndex + rightIndex) / 2
			self.sort(array,leftIndex,mid)
			self.sort(array,mid +1,rightIndex)
			self.__merge(array,leftIndex,rightIndex)
	
	def __merge(self,array,leftIndex,rightIndex):
		mid = (leftIndex + rightIndex) / 2
		leftArray = array[leftIndex:mid +1]
		rightArray = array[mid +1: rightIndex + 1]
		i = 0
		j = 0
		while(i < len(leftArray) and j < len(rightArray)):
			if(leftArray[i] < rightArray[j]):
				array[leftIndex] = leftArray[i]
				i = i  +1
			else:
				array[leftIndex] = rightArray[j]
				j = j + 1
			leftIndex = leftIndex + 1
	
		while(i < len(leftArray)):
			array[leftIndex] = leftArray[i]
			i = i + 1
			leftIndex = leftIndex + 1
		while(j < len(rightArray)):
			array[leftIndex] = rightArray[j]
			j = j + 1
			leftIndex = leftIndex + 1

	def __validateParams(self,array,leftIndex,rightIndex):
		if(array == None or leftIndex < 0 or rightIndex > len(array) - 1 or leftIndex >= rightIndex):
			return False
		else:
			return True

sort = MergeSort()
lista = [9,8,7,6,5,4,3,2,1]
sort.sort(lista,5,8)
print lista

File: Ordination/C++/MergeSort.cpp
#include <iostream>


template<typename T> class MergeSort{
	public:
		void sort(T array[],int leftIndex,int rightIndex,int size){
			if(validateParams(array,leftIndex,rightIndex,size)){
				int mid = (leftIndex + rightIndex)  / 2;
				this->sort(array,leftIndex,mid,size);
				this->sort(array,mid + 1,rightIndex,size);
				this->merge(array,leftIndex,rightIndex,size);
			}else{
				//Do nothing
			}
		}
		void sort(T array[],int size){
			this->sort(array,0,size-  1,size);
			this->showArray(array,size);
		}
	private:
		void merge(T array[],int leftIndex,int rightIndex,int size){
			int mid = (leftIndex + rightIndex) / 2;
			T newArray[size];
			int i = leftIndex;
			int pivot = leftIndex;
			int j = mid + 1;
		
			while(i <= mid && j <= rightIndex){
				if(array[i] < array[j]){
					newArray[pivot++] = array[i++];
				}else{
					newArray[pivot++] = array[j++];
				}
			}
			while(i <= mid){
				newArray[pivot++] = array[i++];
			}
			while(j <= rightIndex){
				newArray[pivot++] = array[j++];
			}
		
			for(int i = leftIndex; i <= rightIndex;i++){
				array[i] = newArray[i];
			}
		}
		
		T* fillArray(T array[],int leftIndex,int rightIndex){
			int size = (rightIndex - leftIndex) + 1;
			if(size > 0){
				T newArray[size];
				int index = 0;
				for(int i = leftIndex;i <= rightIndex;i++){
					newArray[index] = array[i];
					index++;
				}
				return newArray;
			}else{
				T newArray[0];
				return newArray;
			}
		}

		bool validateParams(T array[],int leftIndex,int rightIndex,int size){
			if(array == NULL || leftIndex < 0 || rightIndex > size - 1 || leftIndex >= rightIndex){
				return false;
			}else{
				return true;
			}
		}
		
		void swap(T array[], int i,int j){
			T aux = array[i];
			array[i] = array[j];
			array[j] = aux;
		}
	
		void showArray(T array[],int size){
			std::cout << "Array : [";	
			for(int i = 0 ; i < size; i++){
				if(i != size - 1){
					std::cout << array[i] << ",";
				}else{
					std::cout << array[i] << "]";
				}
			}
			std::cout << std::endl;
		}
	
};
